### PR TITLE
Refactor(pyavd): Refactor backend ANTA code of PyAVD

### PR DESCRIPTION
**Pull request by carl-baillargeon**
## Change Summary

Refactor and reorganize backend ANTA code of PyAVD.

## Component(s) name

PyAVD

## Proposed changes

1- Converted Pydantic models to dataclasses and updated class names to follow our codebase
2- Removed the `conditional_keys` field from `AvdTestSpec` to use the new `skip_if_missing_config` decorator instead
3- Also added the relevant input factories that use the new decorator
4- Removed `ignore_is_deployed` knob from `AvdCatalogGenerationSettings` - users can simply look for the `is_deployed` key prior to calling `get_device_test_catalog`
5- Moved the filtering logic of `AvdTestSpec` to a function in utils to simplify `get_device_test_catalog`
6- Moved the creation of `InputFactoryDataSource` (previously `DeviceTestContext`) to `get_device_test_catalog`
7- Removed perf_counter leftover logs from a previous PR

## How to test

`anta_runner` molecule.

### User Checklist

- Add unit tests

### Repository Checklist

- [X] My code has been rebased from devel before I start
- [X] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [X] My change requires a change to the documentation and documentation have been updated accordingly.
- [X] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/extensions/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: docs/porting-guides/6.x.x.md [hunk 1]
```diff
@@ -1103,22 +1107,104 @@ def get_device_test_catalog(
 ) -> AntaCatalog
 ```
 
-All code using this function must be updated as shown in the following example:
+**Action Required:** Update your code to convert structured configurations into `AvdFabricData` before calling the function.
```
Comment (gmuloc) [inline docs/porting-guides/6.x.x.md:1283]: we may need to sync on this so that we are consistent in the portigng guide entries for these signature changes. Lets keep track in the main umbrella issue

File: docs/porting-guides/6.x.x.md [hunk 2]
```diff
@@ -1131,36 +1131,160 @@ tenants:
...
 
 ```diff
-def get_device_test_catalog(
-    hostname: str,
-    structured_config: dict,
--   minimal_structured_configs: dict[str, MinimalStructuredConfig],
-+   fabric_data: AvdFabricData,
-    settings: AvdCatalogGenerationSettings | None = None,
-) -> AntaCatalog
+  def get_device_test_catalog(
```
Comment (MaheshGSLAB) [inline docs/porting-guides/6.x.x.md:1274]: is this space change expected?

**Review by MaheshGSLAB — APPROVED** [review]
LGTM

File: docs/porting-guides/6.x.x.md (hunk 2 above)
Comment (carl-baillargeon) [reply docs/porting-guides/6.x.x.md:1274]: Yes.

File: docs/porting-guides/6.x.x.md [hunk 3]
```diff
@@ -1188,36 +1188,160 @@ Enable the execution of custom tests by explicitly setting `user_catalogs_enable
...
+  from pyavd.api._anta import AvdFabricData
+
+  dc1_leaf1a_structured_config = {"hostname": "DC1-LEAF1A", "metadata": {"is_deployed": False}}
+
+  structured_configs = {"DC1-LEAF1A": dc1_leaf1a_structured_config}
+  fabric_data = AvdFabricData.from_structured_configs(structured_configs)
+
+- anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data)
++ if dc1_leaf1a_structured_config.get("metadata", {}).get("is_deployed", True) is False:
++     anta_catalog = AntaCatalog()
```
Comment (gmuloc) [inline docs/porting-guides/6.x.x.md]: do we want to suggest generating empty catalogs?

File: docs/porting-guides/6.x.x.md [hunk 4]
```diff
@@ -1188,36 +1188,160 @@ Enable the execution of custom tests by explicitly setting `user_catalogs_enable
...
++ else:
++     anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data)
+```
+
+#### 3. Removal of `input_factory_settings` Setting
+
+The `AvdCatalogGenerationSettings.input_factory_settings` attribute, previously used to globally enable BGP validation for VRF peers, has been removed. Control has moved to the device structured configuration.
+
+**Action Required:** Instead of passing `InputFactorySettings(allow_bgp_vrfs=True)`, you must now set `metadata.validate_state` to `True` on specific BGP neighbors or neighbor interfaces (RFC 5549) within the structured configuration.
```
Comment (gmuloc) [inline docs/porting-guides/6.x.x.md]: We had a debate with Claus this morning on another PR and decided to remove the space so editing for alignment

```suggestion
**Action Required:** Instead of passing `InputFactorySettings(allow_bgp_vrfs=True)`, you must now set `metadata.validate_state` to `True` on specific BGP neighbors or neighbor interfaces (RFC5549) within the structured configuration.
```

File: docs/porting-guides/6.x.x.md [hunk 5]
```diff
@@ -1188,36 +1188,160 @@ Enable the execution of custom tests by explicitly setting `user_catalogs_enable
...
++     }
++ }
+
+  structured_configs = {"DC1-LEAF1A": dc1_leaf1a_structured_config}
+  fabric_data = AvdFabricData.from_structured_configs(structured_configs)
+
+- input_factory_settings = InputFactorySettings(allow_bgp_vrfs=True)
+- settings = AvdCatalogGenerationSettings(input_factory_settings=input_factory_settings)
+- anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data, settings=settings)
 + anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data)
```
Comment (gmuloc) [inline docs/porting-guides/6.x.x.md]: This may be clearer than the long list
```suggestion
  anta_catalog = get_device_test_catalog(
      hostname="DC1-LEAF1A",
      structured_config=dc1_leaf1a_structured_config,
      fabric_data=fabric_data,
-     settings=settings
  )
```

File: docs/porting-guides/6.x.x.md [hunk 6]
```diff
@@ -1188,36 +1188,160 @@ Enable the execution of custom tests by explicitly setting `user_catalogs_enable
...
++ )
+
+  settings = AvdCatalogGenerationSettings(custom_test_specs=(test_spec,))
+```
+
+#### 5. `AvdCatalogGenerationSettings` is now a Dataclass
+
+`AvdCatalogGenerationSettings` has been converted from a Pydantic model to a standard Python `dataclass`.
+
+**Action Required:** If you code relies on Pydantic-specific methods (like validation error handling or serialization), updates are required.
```
Comment (gmuloc) [inline docs/porting-guides/6.x.x.md]:
```suggestion
**Action Required:** If your code relies on Pydantic-specific methods (like validation error handling or serialization), updates are required.
```

File: python-avd/pyavd/_anta/index.py [hunk 7]
```diff
@@ -7,169 +7,52 @@
...
+    AvdTestSpec(test_class=VerifyTransceiversTemperature, input_factory=VerifyTransceiversTemperatureInputFactory),
+    AvdTestSpec(test_class=VerifyVxlanConfigSanity, input_factory=VerifyVxlanConfigSanityInputFactory),
+    AvdTestSpec(test_class=VerifyZeroTouch),
 ]
 """List of all ANTA tests with their specifications that AVD will run by default."""
 
 AVD_TEST_INDEX.sort(key=lambda x: x.test_class.name)
 """Sort the test index by the test class name."""
-
-AVD_TEST_NAMES: list[str] = [test.test_class.name for test in AVD_TEST_INDEX]
```
Comment (gmuloc) [inline python-avd/pyavd/_anta/index.py:174]: Note to self: Not needed anymore?

File: python-avd/pyavd/_anta/input_factories/_decorators.py
Comment (gmuloc) [inline python-avd/pyavd/_anta/input_factories/_decorators.py:1]: does it need to be _decorators when it is already under _anta?

File: python-avd/pyavd/_anta/input_factories/_decorators.py [hunk 8]
```diff
@@ -46,9 +48,27 @@ def skip_if_wan_router(func: Callable[[F], Iterator[R]]) -> Callable[[F], Iterat
...
-        if self.device.is_wan_router:
+        if self.data_source.is_wan_router:
             self.logger_adapter.debug(LogMessage.DEVICE_IS_WAN_ROUTER)
             return
         yield from func(self)
 
     return wrapper
+
+
+def skip_if_missing_config(*keys: StructuredConfigKey) -> Callable[[Callable[[F], Iterator[R]]], Callable[[F], Iterator[R]]]:
```
Comment (gmuloc) [inline python-avd/pyavd/_anta/input_factories/_decorators.py]: it may be more readable to do something like

```
T_AntaTestInputFactoryCreate = Callable[[F], Iterator[R]]
```

File: python-avd/pyavd/_anta/input_factories/_decorators.py [hunk 9]
```diff
@@ -46,9 +48,27 @@ def skip_if_wan_router(func: Callable[[F], Iterator[R]]) -> Callable[[F], Iterat
...
+        @wraps(func)
+        def wrapper(self: F) -> Iterator[R]:
+            # Check if all keys resolve to a truthy value in the config
+            if not all(get_v2(self.structured_config, key.value) for key in keys):
+                formatted_keys = StructuredConfigKey.to_string_list(list(keys))
+                self.logger_adapter.debug(LogMessage.INPUT_NO_DATA_MODELS, data_models=", ".join(formatted_keys))
+                return
+            yield from func(self)
+
+        return wrapper
```
Comment (gmuloc) [inline python-avd/pyavd/_anta/input_factories/_decorators.py:75]: slight perf improvement in case the decorated function is called multiple times.

```suggestion
    key_values = StructuredConfigKey.to_string_list(keys)
    # or simply
    # key_values = [k.value for k in keys]
    
    def decorator(func: Callable[[F], Iterator[R]]) -> Callable[[F], Iterator[R]]:
        @wraps(func)
        def wrapper(self: F) -> Iterator[R]:
            # Check if all keys resolve to a truthy value in the config
            if not all(get_v2(self.structured_config, value) for value in key_values):
                self.logger_adapter.debug(LogMessage.INPUT_NO_DATA_MODELS, data_models=", ".join(key_values))
                return
            yield from func(self)

        return wrapper
```

File: python-avd/pyavd/_anta/input_factories/avt.py [hunk 10]
```diff
@@ -26,6 +28,7 @@ class VerifyAVTSpecificPathInputFactory(AntaTestInputFactory[VerifyAVTSpecificPa
     `router_path_selection.path_groups.static_peers`.
     """
 
+    @skip_if_missing_config(StructuredConfigKey.ROUTER_AVT, StructuredConfigKey.ROUTER_PATH_SELECTION)
```
Comment (gmuloc) [inline python-avd/pyavd/_anta/input_factories/avt.py:31]: not for this PR but this `StructuredConfigKey`, is it because of some linter to make it happy by note repeating strings?

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py [hunk 11]
```diff
@@ -3,52 +3,73 @@
...
+    """Whether to include extra fabric-wide validation tests in the catalog."""
+
+    def __post_init__(self) -> None:
+        """Validate the `output_dir` attribute if provided."""
+        if self.output_dir is None:
+            return
+        path = Path(self.output_dir)
         if not (path.exists() and path.is_dir()):
-            msg = f"Provided output_dir {value} does not exist or is not a directory"
+            msg = f"Provided output_dir {self.output_dir} does not exist or is not a directory"
```
Comment (gmuloc) [inline python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]:
```suggestion
            msg = f"Provided output_dir {self.output_dir} does not exist or is not a directory."
```

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py [hunk 12]
```diff
@@ -3,52 +3,73 @@
...
             raise ValueError(msg)
-        return path
+
+    def get_filtered_test_specs(self, avd_test_specs: list[AvdTestSpec]) -> list[AvdTestSpec]:
```
Comment (gmuloc) [inline python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]: do you need to parameterize this? what if avd_test_specs contains invalid names?

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py [hunk 13]
```diff
@@ -3,52 +3,73 @@
...
+            if name in self.skip_tests:
+                continue
+
+            # If remaining_run_tests is specified, only include tests in that set
+            if remaining_run_tests and name not in remaining_run_tests:
+                continue
+
+            final_test_specs.append(test)
+
+        # Add custom test specs, avoiding duplicates
```
Comment (gmuloc) [inline python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]: duplicates in custom_test_specs won't be avoided with

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py [hunk 14]
```diff
@@ -3,52 +3,73 @@
...
+        final_test_specs.extend([test for test in self.custom_test_specs if test not in final_test_specs])
```
Comment (gmuloc) [inline python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]:
```suggestion
        final_test_specs.extend(test for test in self.custom_test_specs if test not in final_test_specs)
```

File: python-avd/pyavd/api/_anta/avd_test_spec.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        # Check if the test class has an `Input` model and if it has required fields
+        if "Input" in self.test_class.__dict__ and issubclass((input_class := self.test_class.__dict__["Input"]), AntaTest.Input):
+            for f_name, f_info in input_class.model_fields.items():
+                # No need to check the base class fields
+                if f_name in {"result_overwrite", "filters"}:
+                    continue
+                # If a required field is present, an input factory must be provided
+                if f_info.is_required() and self.input_factory is None:
+                    msg = f"AvdTestSpec for {self.test_class.name} must have `input_factory`"
+                    raise ValueError(msg)
```
Comment (gmuloc) [inline python-avd/pyavd/api/_anta/avd_test_spec.py]:
```suggestion
        input_class = getattr(self.test_class, "Input", None)
        if not (isinstance(input_class, type) and issubclass(input_class, AntaTest.Input)):
            return

        excluded_fields = {"result_overwrite", "filters"}        
        requires_input = any(
            f_info.is_required()
            for f_name, f_info in input_class.model_fields.items()
            if f_name not in excluded_fields
        )

        if requires_input and self.input_factory is None:
            msg = f"AvdTestSpec for {self.test_class.name} must have `input_factory`."
            raise ValueError(msg)
```

File: python-avd/pyavd/get_device_test_catalog.py [hunk 16]
```diff
@@ -47,59 +44,26 @@ def get_device_test_catalog(
     Returns:
         The generated ANTA catalog for the device.
     """
+    from dataclasses import asdict  # noqa: PLC0415
```
Comment (gmuloc) [inline python-avd/pyavd/get_device_test_catalog.py]: this is a builtin so I think you can put it as a normal import

File: python-avd/pyavd/get_device_test_catalog.py [hunk 17]
```diff
@@ -47,59 +44,26 @@ def get_device_test_catalog(
...
-    # Add custom test specs, avoiding duplicates
-    filtered_test_specs.extend([test for test in settings.custom_test_specs if test not in filtered_test_specs])
-
-    catalog = create_catalog(hostname, structured_config, fabric_data, settings, filtered_test_specs)
+    catalog = create_catalog(data_source, final_test_specs)
 
     if settings.output_dir:
         dump_anta_catalog(hostname, catalog, settings.output_dir)
 
-    stop_time = perf_counter()
```
Comment (gmuloc) [inline python-avd/pyavd/get_device_test_catalog.py:102]: we dump the perf?

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+    """No run_tests or skip_tests provided. Should return all."""
+    settings = AvdCatalogGenerationSettings()
+    result = settings.get_filtered_test_specs(AVD_TEST_INDEX)
+    assert result == AVD_TEST_INDEX
+
+
+def test_get_filtered_test_specs_invalid_test_names_in_run() -> None:
+    """Invalid name in run_tests should raise ValueError."""
+    settings = AvdCatalogGenerationSettings(run_tests=("VerifyReachability", "InvalidTest1", "InvalidTest2"))
+    with pytest.raises(ValueError, match="Invalid test names in run_tests or skip_tests filters: InvalidTest1, InvalidTest2"):
```
Comment (gmuloc) [inline python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]:
```suggestion
    with pytest.raises(ValueError, match="Invalid test name(s) in 'run_tests' or 'skip_tests' filters: InvalidTest1, InvalidTest2"):
```

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    # User asks to run A and B, but explicitly skips B. Result should be A only.
+    run_tests = ("VerifyNTP", "VerifyReachability")
+    skip_tests = ("VerifyReachability",)
+    settings = AvdCatalogGenerationSettings(run_tests=run_tests, skip_tests=skip_tests)
+    result = settings.get_filtered_test_specs(AVD_TEST_INDEX)
+    result_names = [s.test_class.name for s in result]
+    assert result_names == ["VerifyNTP"]
+
+
+def test_get_filtered_test_specs_both_filters_overlap_no_remaining_run_tests() -> None:
```
Comment (gmuloc) [inline python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: I would argue this behavior is quite weird. 
I would expect no test is run if I ask for ONLY a specific test and I skip it.
I know this comes from your logic of sets but from a user standpoint you are running stuff I asked you not to run when I limited the ran tests with `run_tests`

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+    custom_test_specs = (test_spec,)
+    settings = AvdCatalogGenerationSettings(custom_test_specs=custom_test_specs)
+    result = settings.get_filtered_test_specs(AVD_TEST_INDEX)
+    assert test_spec in result
+    assert len(result) == len(AVD_TEST_INDEX) + len(custom_test_specs)
+
+
+def test_get_filtered_test_specs_custom_test_specs_deduplication() -> None:
+    """If a custom test is already in the list (value equality), it is not added twice."""
+    test_spec = AvdTestSpec(test_class=VerifyReachability, input_factory=VerifyReachabilityInputFactory)
```
Comment (gmuloc) [inline python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: if you were to put twice the same test class not in AVD_TEST_INDEX you would end up with a duplicate

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+    custom_test_specs = (test_spec,)
+    settings = AvdCatalogGenerationSettings(custom_test_specs=custom_test_specs)
+    result = settings.get_filtered_test_specs(AVD_TEST_INDEX)
+    verify_reachability_count = sum(1 for s in result if s.test_class.name == test_spec.test_class.name)
+    assert verify_reachability_count == 1
+    assert result == AVD_TEST_INDEX
+
+
+def test_get_filtered_test_specs_custom_test_specs_with_filters() -> None:
```
Comment (gmuloc) [inline python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: you should add a test where I try to skip a test in the custom_test_specs and make sure it stays in there (as ignored by the skip)

also another one would be to actually do

```
custom_test = VerifyReachability
skip_test = VerifyReachability
```

and making sure that you still have VerifyReachability from custom in the result

**Review by gmuloc — CHANGES_REQUESTED** [review]

File: docs/porting-guides/6.x.x.md (hunk 1 above)
Comment (carl-baillargeon) [reply docs/porting-guides/6.x.x.md:1283]: Agreed. As discussed live we will sync during documentation week :)

File: docs/porting-guides/6.x.x.md (hunk 3 above)
Comment (carl-baillargeon) [reply docs/porting-guides/6.x.x.md]: Nop. Fixed in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

File: docs/porting-guides/6.x.x.md (hunk 4 above)
Comment (carl-baillargeon) [reply docs/porting-guides/6.x.x.md]: All good. Fixed in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

File: docs/porting-guides/6.x.x.md (hunk 5 above)
Comment (carl-baillargeon) [reply docs/porting-guides/6.x.x.md]: Nice. I adopted this style in other places as well in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

File: docs/porting-guides/6.x.x.md (hunk 6 above)
Comment (carl-baillargeon) [reply docs/porting-guides/6.x.x.md]: Fixed in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

File: python-avd/pyavd/_anta/input_factories/avt.py (hunk 10 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_anta/input_factories/avt.py:31]: Not really. Using enums in case our data models change and to have them defined in a central place.

File: python-avd/pyavd/api/_anta/avd_test_spec.py (hunk 15 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/api/_anta/avd_test_spec.py]: This won't work because `input_class` will fallback to `AntaTest.Input` and never hit the first condition:
```
>>> input_class = getattr(VerifyBGPPeerSession, "Input", None)
>>> input_class
<class 'anta.tests.routing.bgp.VerifyBGPPeerSession.Input'>
>>> isinstance(input_class, type) and issubclass(input_class, AntaTest.Input)
True
>>> input_class = getattr(VerifyOSPFNeighborState, "Input", None)
>>> input_class
<class 'anta.models.AntaTest.Input'>
>>> isinstance(input_class, type) and issubclass(input_class, AntaTest.Input)
True
>>>
```

File: python-avd/pyavd/get_device_test_catalog.py (hunk 17 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/get_device_test_catalog.py:102]: Yes I forgot to remove this in the inital implementation. We can benchmark in the calling code if needed or codspeed later :)

File: python-avd/pyavd/_anta/index.py (hunk 7 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_anta/index.py:174]: Nop.

File: python-avd/pyavd/_anta/input_factories/_decorators.py
Comment (carl-baillargeon) [reply python-avd/pyavd/_anta/input_factories/_decorators.py:1]: I will rename modules and classes (AVDXxx) in a separate PR without any other changes to make it easier to review.

File: python-avd/pyavd/_anta/input_factories/_decorators.py (hunk 8 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_anta/input_factories/_decorators.py]: Fixed.

File: python-avd/pyavd/_anta/input_factories/_decorators.py (hunk 9 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/_anta/input_factories/_decorators.py:75]: Applied your suggestion and removed unused `StructuredConfigKey.to_string_list`.

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py (hunk 11 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]: Fixed.

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py (hunk 12 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]: Good call. ~~I removed the `avd_test_specs` param and used `AVD_TEST_INDEX` in the function directly.~~

That did not work because of circular imports. I decided to move this method out and put it in `_anta/utils.py` to avoid this. It makes more sense to put it there besides the index anyways.

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py (hunk 13 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]: I decided to remove the deduplication check altogether to honor user intent. If they add duplicates, too bad.

File: python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py (hunk 14 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/api/_anta/avd_catalog_generation_settings.py]: No longer deduplicating so no longer needed.

File: python-avd/pyavd/api/_anta/avd_test_spec.py (hunk 15 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/api/_anta/avd_test_spec.py]: I refactored following your suggestion but keeping the check with `__dict__`.

File: python-avd/pyavd/get_device_test_catalog.py (hunk 16 above)
Comment (carl-baillargeon) [reply python-avd/pyavd/get_device_test_catalog.py]: Moved this out.

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py (hunk 18 above)
Comment (carl-baillargeon) [reply python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: Done.

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py (hunk 19 above)
Comment (carl-baillargeon) [reply python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: Ah yes good catch. This is actually a bug in the current implementation (and in molecule). I changed the code to handle this corner case (a single test in both filters). Updated unit tests. Molecule updated as well.

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py (hunk 20 above)
Comment (carl-baillargeon) [reply python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: Removed the deduplication in the code so no longer needed.

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py (hunk 21 above)
Comment (carl-baillargeon) [reply python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: The first suggestion is not possible since we cannot skip test names that are not in the default test index :) I added the second test suggestion.

File: python-avd/pyavd/_anta/input_factories/avt.py (hunk 10 above)
Comment (gmuloc) [reply python-avd/pyavd/_anta/input_factories/avt.py:31]: lets live with this for now :)

File: python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py (hunk 21 above)
Comment (gmuloc) [reply python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: the first suggestion is possible if I add a custom test that is in the default test index

Comment (gmuloc) [reply python-avd/tests/pyavd/api/_anta/test_avd_catalog_generation_settings.py]: which is my second option - so I guess both were saying the same stuff in teh end

**Review by gmuloc — APPROVED** [review]
